artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/ProtonServerSenderContext.java

```diff
@@ -584,56 +583,45 @@ void deliver() {
             deliveryAnnotationsToEncode = null;
          }
 
-         LargeBodyReader context = message.getLargeBodyReader();
          try {
-            context.open();
-            try {
+            final ByteBuf frameBuffer = PooledByteBufAllocator.DEFAULT.directBuffer(frameSize, frameSize);
+            final NettyReadable frameView = new NettyReadable(frameBuffer);
+            try (LargeBodyReader context = message.getLargeBodyReader()) {
+               context.open();
                context.position(position);
                long bodySize = context.getSize();
+               // materialize it so we can use its internal NIO buffer
+               frameBuffer.ensureWritable(frameSize);
 
-               ByteBuffer buf = ByteBuffer.allocate(frameSize);
+               if (position == 0 && sender.getLocalState() != EndpointState.CLOSED && position < bodySize) {
```

Comment:
> "position = 0..." is a pretty awkward way of saying 'have we finished doing the required initial packet work before'. Adding a specific boolean for this might be better

Agree but I understand why @clebertsuconic used position here, to save adding another field member (ie another state variable to preserve/maintain) to track something that position validation could track (more on this later).

> Also it doesnt necessarily actually matter if position <bodySize() 

Good point, this is not a useful check there

> e.g there are no delivery-annotations or header in the original message to skip past, or the subsequently generated ones precisely fit in the buffer with room for nothing else, or the connection was flow controlled before sending anything else

If there are no delivery annotations or header, the deliveryInitialPacket takes care filling the rest of the frame with some file content (and that should be at least one byte, but this is a missing check yet, indeed), increasing position over 0.
If flow control kick in, it would happen before doing it, hence only close or available Netty outbound buffer would resume it (with position still == 0).
Assuming that files contains at least a single byte, using position == 0 should be safe, logically, despite not very intuitive.

Reply:
> > e.g there are no delivery-annotations or header in the original message to skip past, or the subsequently generated ones precisely fit in the buffer with room for nothing else, or the connection was flow controlled before sending anything else
> 
> If there are no delivery annotations or header, the deliveryInitialPacket takes care filling the rest of the frame with some file content (and that should be at least one byte, but this is a missing check yet, indeed), increasing position over 0.

True, forgot that.

> If flow control kick in, it would happen before doing it, hence only close or available Netty outbound buffer would resume it (with position still == 0).
> Assuming that files contains at least a single byte, using position == 0 should be safe, logically, despite not very intuitive.

I meant flow control after sending the initial packet, in the loop. There are 2 routes that it is possible (but unlikely, I admit) to get into the loop without incrementing the position field, which would mean it could go through the initial stuff again and generate the initial packet again if it flow controls.